src/D2L.CodeStyle.Analyzers/Immutability/ImmutableGenericArgument.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	public static class ImmutableGenericArgument {
+		public static bool InterfacesDemandImmutability(
+			INamedTypeSymbol symbol,
+			ITypeSymbol argument
+		) {
+			foreach( ITypeSymbol intf in symbol.Interfaces ) {
+				if( SymbolDemandsImmutability(
+					intf,
+					argument
+				) ) {
+					return true;
+				}
+			}
+
+			return false;
+		}
+
+		public static bool ConstraintsDemandImmutabliity(
+			INamedTypeSymbol symbol,
+			ITypeSymbol argument
+		) {
+			int argumentOrdinal = symbol.IndexOfArgument( argument.Name );
+			if (argumentOrdinal < 0) {
+				return false;
+			}
+
+			ITypeParameterSymbol argumentTypeParameter =
+				symbol.TypeParameters[argumentOrdinal];
+
+			if( argumentTypeParameter == default ) {

```

Comment:
I feel like with the ordinal check you have above, this check isn't needed anymore.  I'd be super surprised if Roslyn had a null element in the array